fem/tmop_pa_h2s.cpp

```diff
@@ -0,0 +1,247 @@
+
+   for (int i = 0; i < DIM; i++)
+   {
+      for (int j = 0; j < DIM; j++)
+      {
+         ConstDeviceMatrix ddi1(ie.Get_ddI1(i,j),DIM,DIM);
+         ConstDeviceMatrix ddi2(ie.Get_ddI2(i,j),DIM,DIM);
+         for (int r = 0; r < DIM; r++)
+         {
+            for (int c = 0; c < DIM; c++)
+            {
+               H(r,c,i,j,qx,qy,e) =
+                  weight * (1.0 + c1) * ddi1(r,c)
+                  - c3 * ddi2(r,c)
+                  - c2 * ( di1(i,j) * di2(r,c) + di2(i,j) * di1(r,c) )
+                  + 2.0 * c1 * c3 * di2(r,c) * di2(i,j);
+            }
+         }
+      }
+   }
+}
+
+static MFEM_HOST_DEVICE inline
+void EvalH_077(const int e, const int qx, const int qy,
+               const double weight, const double *Jpt,
+               DeviceTensor<7,double> H)
+{
+   constexpr int DIM = 2;
+   double dI2[4], dI2b[4], ddI2[4];
+   kernels::InvariantsEvaluator2D ie(Args()
+                                     .J(Jpt)
+                                     .dI2(dI2)
+                                     .dI2b(dI2b)
+                                     .ddI2(ddI2));
+   const double I2 = ie.Get_I2(), I2inv_sq = 1.0 / (I2 * I2);
+   ConstDeviceMatrix di2(ie.Get_dI2(),DIM,DIM);
+   for (int i = 0; i < DIM; i++)
+   {
+      for (int j = 0; j < DIM; j++)
+      {
+         ConstDeviceMatrix ddi2(ie.Get_ddI2(i,j),DIM,DIM);
+         for (int r = 0; r < DIM; r++)
+         {
+            for (int c = 0; c < DIM; c++)
+            {
+               H(r,c,i,j,qx,qy,e) =
+                  weight * 0.5 * (1.0 - I2inv_sq) * ddi2(r,c)
+                  + weight * (I2inv_sq / I2) * di2(r,c) * di2(i,j);
+            }
+         }
+      }
+   }
+}
+
+MFEM_REGISTER_TMOP_KERNELS(void, SetupGradPA_2D,
+                           const Vector &x_,
+                           const double metric_normal,
+                           const int mid,
+                           const int NE,
+                           const Array<double> &w_,
+                           const Array<double> &b_,
+                           const Array<double> &g_,
+                           const DenseTensor &j_,
+                           Vector &h_,
+                           const int d1d,
+                           const int q1d)
+{
+   MFEM_VERIFY(mid == 1 || mid == 2 || mid == 7 || mid == 77,
+               "Metric not yet implemented!");
+
+   constexpr int DIM = 2;
+   constexpr int NBZ = 1;
+   const int D1D = T_D1D ? T_D1D : d1d;
+   const int Q1D = T_Q1D ? T_Q1D : q1d;
+
+   const auto W = Reshape(w_.Read(), Q1D, Q1D);
+   const auto b = Reshape(b_.Read(), Q1D, D1D);
+   const auto g = Reshape(g_.Read(), Q1D, D1D);
+   const auto J = Reshape(j_.Read(), DIM, DIM, Q1D, Q1D, NE);
+   const auto X = Reshape(x_.Read(), D1D, D1D, DIM, NE);
+   auto H = Reshape(h_.Write(), DIM, DIM, DIM, DIM, Q1D, Q1D, NE);
+
+   MFEM_FORALL_2D(e, NE, Q1D, Q1D, NBZ,
+   {
+      const int D1D = T_D1D ? T_D1D : d1d;
```

Comment:
This could make the code a bit more clear:
```suggestion
      using kernels;
      const int D1D = T_D1D ? T_D1D : d1d;
```